### PR TITLE
feat: add Two factor auth feature

### PR DESCRIPTION
**Pull request by G3root**
simple demo of the auth flow

[Screencast 2023-11-10 07:57:50.webm](https://github.com/documenso/documenso/assets/84864519/45be70a4-ba22-4329-a847-8beaa2866f3f)

### CODE REVIEW

**Comment by Mythie** [comment]
Looks like this could be huge Nafees! I'll review personally but may need some time to do so 🙌🏻

File: packages/lib/universal/crypto.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  data: string;
+  key: string;
+};
+
+export const getEncryptionKey = () => {
+  const encryptionKey = DOCUMENSO_ENCRYPTION_KEY;
+
+  if (!encryptionKey) {
+    throw new Error('Missing encryption key');
+  }
```
Comment (catalinpit) [inline packages/lib/universal/crypto.ts:20]: Space please

File: packages/lib/server-only/2fa/setup-2fa.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      message: ErrorCode.INCORRECT_IDENTITY_PROVIDER,
+    });
+  }
+
+  if (!user.password) {
+    throw new TRPCError({
+      code: 'INTERNAL_SERVER_ERROR',
+      message: ErrorCode.USER_MISSING_PASSWORD,
+    });
+  }
```
Comment (catalinpit) [inline packages/lib/server-only/2fa/setup-2fa.ts:48]: Space please

File: packages/lib/server-only/2fa/enable-2fa.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  await prisma.user.update({
+    where: {
+      id: user.id,
+    },
+    data: {
+      twoFactorEnabled: true,
+    },
+  });
+
+  return { message: '2fa enabled successfully' };
```
Comment (catalinpit) [inline packages/lib/server-only/2fa/enable-2fa.ts:66]: "2FA" please, and maybe "!" at the end

File: packages/lib/server-only/2fa/enable-2fa.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+type enableTwoFactorAuthenticationOptions = {
+  user: User;
+  code: string;
+};
+
+export const enableTwoFactorAuthentication = async ({
+  user,
+  code,
+}: enableTwoFactorAuthenticationOptions) => {
+  const encryptionKey = process.env.DOCUMENSO_ENCRYPTION_KEY;
```
Comment (catalinpit) [inline packages/lib/server-only/2fa/enable-2fa.ts:18]: Lines 18 to 32 have been re-used in 2 places already. If it is used in other places as well, should we extract it, or?

File: packages/lib/server-only/2fa/disable-2fa.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      id: user.id,
+    },
+    data: {
+      twoFactorEnabled: false,
+      twoFactorBackupCodes: null,
+      twoFactorSecret: null,
+    },
+  });
+
+  return { message: '2fa disabled successfully' };
```
Comment (catalinpit) [inline packages/lib/server-only/2fa/disable-2fa.ts:60]: Same comment like the 2FA enable message

File: packages/lib/server-only/2fa/authenticate-2fa.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  if (!user.twoFactorSecret) {
+    throw new Error(ErrorCode.TWO_FACTOR_MISSING_SECRET);
+  }
+
+  if (user.identityProvider !== 'DOCUMENSO') {
+    throw new Error(ErrorCode.INCORRECT_IDENTITY_PROVIDER);
+  }
+
+  if (totpCode) {
```
Comment (catalinpit) [inline packages/lib/server-only/2fa/authenticate-2fa.ts:34]: Couldn't this whole block be simplified a little bit?

File: apps/web/src/components/forms/2fa/authenticator-app-disable-dialog.tsx [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+import { useToast } from '@documenso/ui/primitives/use-toast';
+
+type AuthenticatorAppDisableDialogProps = {
+  open: boolean;
+  onOpenChange: (_val: boolean) => void;
+  onDisabled: () => void;
+};
+
+const formSchema = z.object({
+  password: z.string().min(6),
```
Comment (catalinpit) [inline apps/web/src/components/forms/2fa/authenticator-app-disable-dialog.tsx:38]: Custom error messages for validation stuff would be really useful so users don't get the generic & meaningful ones. This comment applies to all Zod schemas.

File: apps/web/src/components/forms/2fa/authenticator-app-setup-dialog.tsx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        variant: 'default',
+      });
+    },
+  });
+
+  const onSubmit = async (data: TFormSchema) => {
+    try {
+      await enable(data);
+    } catch (error) {
+      return;
```
Comment (catalinpit) [inline apps/web/src/components/forms/2fa/authenticator-app-setup-dialog.tsx:84]: No error message returned?

File: apps/web/src/components/forms/2fa/authenticator-app.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  const [modalState, setModalState] = useState<ModalState>(null);
+  const [data, setData] = useState<Awaited<ReturnType<typeof setup>> | null>(null);
+
+  const handleVerified = async (password: string) => {
+    try {
+      const data = await setup({ password });
+      setData(data);
+      setModalState('setup-2fa');
+    } catch (error) {
+      console.error(error);
```
Comment (catalinpit) [inline apps/web/src/components/forms/2fa/authenticator-app.tsx:35]: Is `console.error` the best way to handle the error?

**Comment by Mythie** [comment]
It's me! Reviewing now 😎

**Comment by Mythie** [comment]
Honestly that was some seriously awesome work! 🚀 

I've just got some merge conflicts and a tiny bit more tidying of code I added to fix up and it's good to go 😄 

/tip 300

**Review by Mythie — APPROVED** [review]

**Pull request merged**